### PR TITLE
[MVT] The test viewer

### PR DESCRIPTION
**Pull request by oterral**
DO NOT MERGE

This PR contains the test viewer.

It's based on these 3 PRs:

- https://github.com/geoadmin/mf-geoadmin3/pull/4493  (OL5 migration)-> soon in master
- https://github.com/geoadmin/mf-geoadmin3/pull/4259 (web mercator)
- https://github.com/geoadmin/mf-geoadmin3/pull/4272 (vector tiles bg)

<jenkins>[Test link](https://mf-geoadmin3.int.bgdi.ch/mvt/index.html)</jenkins>

### CODE REVIEW

**Comment by davidoesch** [comment]
bg switcher contains
- [ ] existing ch.swisstopo.pixelkarte-farbe Color Map Icon
- [ ] existing voidLayer "white background  Icon
- [ ] new  Hybrid Map (Instead of Aerial Imagery)
- [ ] new  Lightbase Map (Instead of Grey map)

**Comment by davidoesch** [comment]
hi @loicgasser @oterral  https://github.com/geoadmin/mf-geoadmin3/files/2534923/RE4_ALPHA_TEST_RE4_layer_styles.xlsx

**Comment by ahocevar** [comment]
It looks like the way the OpenLayers layers are created here adds some unneccessary overhead. I have created a minimal [example](https://codesandbox.io/s/824jn1vrp8) with the same dataset (except for a different hillshade, because the one used in this test app is restricted). If you compare the performance of [your test app](https://mf-geoadmin3.int.bgdi.ch/mvt/index.html) with [mine](https://824jn1vrp8.codesandbox.io/#/center/7.709,46.614/zoom/8), you'll notice that mine feels more responsive.

**Comment by davidoesch** [comment]
significant faster @ahocevar ...  we need to reallylook at this example @oterral @loicgasser  cc @gjn

**Comment by loicgasser** [comment]
@ahocevar we are doing 2 things differently.

1. We use olms.stylefunction to create and apply a style function to a VectorTile Layer
2. We create a layer group with the hillshade and the vector layers

Could this explain the difference? It makes a lot more sense for us to use this approach given the current organization of the layers and their configurations.

**Comment by ahocevar** [comment]
I noticed that `olms.stylefunction` is called four times instead of just once, so maybe you are creating more than one vector tile layer with the same style? Also, do you have many empty vector layers on the map? Then you might be running into openlayers/openlayers#8514.

Note that in ol-mapbox-style, `stylefunction` is the lowest level way to get mapbox styles, without any sprite and font handling. You may want to use `applyStyle` instead. That should have no impact on performance though.

**Comment by oterral** [comment]
As @loicgasser said  we choose this method to keep the same functionalities for all type of layers within the app. When using the full screen example. It looks smoother but I don 't see a "significant" change.

So in the case of map.geo.admin one layer is associated with one gl style.

For the background layers we take the glStyle associated to the bg layer then create a LayerGroup then we apply the same glStyle for all the child, each child corresponds to one source in the glStyle.

So yes we create one ` styleFunction ` for each vector tile layer. 

> Note that in ol-mapbox-style, stylefunction is the lowest level way to get mapbox styles, without any sprite and font handling. You may want to use applyStyle instead. That should have no impact on performance though.

Last time I tested,  ` applyStyle`  didn't work but it was at the beginning of the year. It worth a another try but as you said I don't think it changes the performance. What we also like with `styleFunction`,  it is that we keep control of the sprite loading data, so we can handle CORS/Security/proxy special behavior inherent to Swisstopo.

>I noticed that olms.stylefunction is called four times instead of just once, so maybe you are creating more than one vector tile layer with the same style? Also, do you have many empty vector layers on the map? Then you might be running into openlayers/openlayers#8514.

Yes there is a small problem here it should call it only once for openmaptiles, but the `styleFunction` should be overwritten each time so it shouldn't be a problem. I will check, thx for pointing this out.

**Comment by danduk82** [comment]
@oterral , @loicgasser , the save kml function does not work, it writes the files to public-{staging}-bgdi-ch bucket at the root `/` instead of `/files/`. When you try to download it again you get a `403` error, which is misleading because `S3:ListObjects*` policy is missing, but it should be a `404`. I have used this test file, which should be at `https://public.int.bgdi.ch/files/LBZwOANnTgame8vi5ARtQg` but is at `https://public.int.bgdi.ch/LBZwOANnTgame8vi5ARtQg`

see 
```
aws s3 ls s3://public-int-bgdi-ch/files/LBZwOANnTgame8vi5ARtQg
# (no response)
aws s3 ls s3://public-int-bgdi-ch/LBZwOANnTgame8vi5ARtQg
2018-11-14 15:07:09        474 LBZwOANnTgame8vi5ARtQg
```

**Comment by danduk82** [comment]
Or is it intended that it stays at root of bucket instead of going to `/files`? In fact both solutions are good if it runs.

**Comment by loicgasser** [comment]
yes this is how it is at the moment for KMLs in prod. We save to the REST endpoint /files, but we save in the bucket at the root. For the gl styles we save at /gl-styles for the REST endpoint and also save in the bucket with /gl-styles.

**Comment by danduk82** [comment]
ok

**Comment by loicgasser** [comment]
jenkins test

**Comment by danduk82** [comment]
@oterral , it would be great to see the new style on the test-link, could you please have a look why jenkins is failing when trying to `git merge`? There are some (probably stupid) conflicts with `package-lock.json` .

**Comment by davidoesch** [comment]
@danduk82 

Would it make sense to push the data before to a prod environment
 DATA:
https://vectortiles.geo.admin.ch/mbtiles/ch.swisstopo.leichte-basiskarte.vt/<version>/ -> MBT files (Version V001 increment

https://vectortiles.geo.admin.ch/mbtiles/ ch.swisstopo.leichte-basiskarte.vt/<version>/tileset.json   -> JSON file (tileset json)

STYLE:

https://vectortiles.geo.admin.ch/gl-styles/ch.swisstopo.leichte-basiskarte.vt/current/sprite.json -> sprites JSON

https://vectortiles.geo.admin.ch/gl-styles/ch.swisstopo.leichte-basiskarte.vt/current/sprite.png -> sprites PNG

https://vectortiles.geo.admin.ch/gl-styles/ch.swisstopo.leichte-basiskarte.vt/current/style.json -> style JSON
https://vectortiles.geo.admin.ch/fonts/ > FONTS URL

as described in https://jira.swisstopo.ch/browse/VIB2D_SB-158

so we can even finalize the DI process so @mariokeusen can start with the formalism

**Comment by danduk82** [comment]
@davidoesch : planned for tomorrow with @swissmap 
@oterral :+1:

File: src/js/EditController.js [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+goog.require('ga_layers_service');
+
+(function() {
+
+  var module = angular.module('ga_edit_controller', [
+    'pascalprecht.translate',
+    'ga_layers_service'
+  ]);
+
+  module.controller('GaEditController', function($scope, $translate,
```
Comment (gjn) [inline src/js/EditController.js]: Would prefer more specific name.

File: src/js/MainController.js [hunk 2]
```diff
@@ -81,7 +81,7 @@ goog.require('ga_window_service');
         }),
         logo: false
       });
-
+      window.map = map;
```
Comment (gjn) [inline src/js/MainController.js]: Is this really needed? Don't like wokring via global window object. Map should be passed around as parameter.

File: src/components/map/LayersService.js [hunk 3]
```diff
@@ -275,6 +277,114 @@ goog.require('ga_urlutils_service');
...
+                  ]
+
+                }, {
+                  id: 'labels',
+                  regex: /^labels/,
+                  props: [
+                    ['layout', 'visibility', '{toggle}', 'visible', 'none'],
+                    ['paint', 'text-color', '{color}'],
+                    ['layout', 'text-size', '{size}']
+                  ]
```
[the remark this answers is not in the document]
Comment (oterral) [reply src/components/map/LayersService.js]: Have a look to the readme file I ve added a bit of documentation for that.

File: src/js/MainController.js (hunk 2 above)
Comment (oterral) [reply src/js/MainController.js]: probably added for testing purpose. I remove it.

File: src/js/EditController.js (hunk 1 above)
Comment (oterral) [reply src/js/EditController.js]: Yes I will change to gaEditGlStyle

File: src/components/map/MapUtilsService.js [hunk 4]
```diff
@@ -474,6 +488,127 @@ goog.require('ga_urlutils_service');
...
+            olLayer.setSource(olSource);
+
+            if (data.extent) {
+              // Extent in epsg:4326
+              olLayer.setExtent(ol.proj.transformExtent(data.extent,
+                  ol.proj.get('EPSG:4326'), ol.proj.get('EPSG:3857')));
+            }
+
+          }, function(reason) {
+            $window.console.error('Loading source config failed. Reason: ',
```
[the remark this answers is not in the document]
Comment (oterral) [reply src/components/map/MapUtilsService.js]: All the error message you mention are messages for developer.  We don't warn the user when the layersConfig fails for example . All the requests are expected to work.

File: src/components/ExportGlStyleService.js [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+(function() {
+
+  var module = angular.module('ga_exportglstyle_service', [
+    'ga_glstyle_service',
+    'ga_browsersniffer_service',
+    'pascalprecht.translate'
+  ]);
+
+  /**
+   * This service can be used to export a kml file based on some
```
[the remark this answers is not in the document]
Comment (oterral) [reply src/components/ExportGlStyleService.js]: Yes copy/paste failure

File: src/js/EditController.js (hunk 1 above)
Comment (oterral) [reply src/js/EditController.js]: there is already gaEditGlStyle component

**Comment by davidoesch** [comment]
hi @oterral @loicgasser : the current version is not as performant as https://mf-geoadmin3.int.bgdi.ch/teo_opti_ios/index.html 
Did you merge this behaviour?

**Comment by oterral** [comment]
@davidoesch it has been merged, only difference is I put a cheSize of 20 instead of 0 to have interim tiles when you zoom. Otherwise vector tile disappear and reappear each time you zoom.

**Comment by davidoesch** [comment]
dos not work at all on edge
https://mf-geoadmin3.int.bgdi.ch/mvt/index.html?lang=en&topic=ech&bgLayer=ch.swisstopo.leichte-basiskarte.vt&layers=ch.swisstopo.zeitreihen,ch.bfs.gebaeude_wohnungs_register,ch.bav.haltestellen-oev,ch.swisstopo.swisstlm3d-wanderwege&layers_visibility=false,false,false,false&layers_timestamp=18641231,,,&lon=7.49922&lat=46.94335&zoom=11

**Comment by loicgasser** [comment]
These problems seem related to the ol update. Shall we revert the last PR?

**Comment by oterral** [comment]
@davidoesch @loicgasser I think the update of ol wasn't deployed by jenkins, the build failed since ol pr has been merged. Can you retest now ?

**Comment by gjn** [comment]
I also see that https://github.com/geoadmin/config-tileservergl/issues/26 is not yet in int/prod. When can we expect this?

File: src/components/map/LayersService.js [hunk 6]
```diff
@@ -275,6 +277,114 @@ goog.require('ga_urlutils_service');
...
+                serverLayerName: 'ch.swisstopo.leichte-basiskarte.vt',
+                attribution: 'OpenMapTiles, OpenStreetMap contributors,' +
+                    ' swisstopo',
+                subLayersIds: [
+                  // 'OpenMapTiles',
+                  'ch.swisstopo.swissalti3d-reliefschattierung',
+                  'ch.swisstopo.vektorkarte.vt',
+                  'ch.bav.haltestellen-oev.vt',
+                  'ch.swisstopo.amtliches-strassenverzeichnis_validiert',
+                  'ch.swisstopo.swissnames3d.vt'
```
[the remark this answers is not in the document]
Comment (oterral) [reply src/components/map/LayersService.js]: yes with https://github.com/geoadmin/mf-geoadmin3/pull/4657 subLayersIds is not necessary anymore (only useful for debug purpose)

File: src/components/edit/EditDirective.js [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+    'ga_mvt_service',
+    'ga_background_service',
+    'ga_urlutils_service',
+    'ga_layers_service'
+  ]);
+
+  /**
+   * This directive add an interface where you can modify a glStyle.
+   */
+  module.directive('gaEdit', function($rootScope, $window, $translate, gaMvt,
```
[the remark this answers is not in the document]
Comment (boecklic) [reply src/components/edit/EditDirective.js]: -> gaEditStyle

File: src/components/edit/SizeDirective.js [hunk 8]
```diff
@@ -0,0 +1,64 @@
+goog.provide('ga_size_directive');
+
+(function() {
+
+  var module = angular.module('ga_size_directive', []);
+
+  /**
+   * This directive add an interface where you can modify a glStyle.
+   */
+  module.directive('gaSize', function($window) {
```
Comment (boecklic) [inline src/components/edit/SizeDirective.js]: should be more explicit (also comment). this allows to change the width of e.g. streets. -> gaElementWidth

File: src/components/edit/ToggleDirective.js [hunk 9]
```diff
@@ -0,0 +1,50 @@
+goog.provide('ga_toggle_directive');
+
+(function() {
+
+  var module = angular.module('ga_toggle_directive', []);
+
+  /**
+   * This directive add an interface where you can modify a glStyle.
+   */
+  module.directive('gaToggle', function($window) {
```
[the remark this answers is not in the document]
Comment (boecklic) [reply src/components/edit/ToggleDirective.js]: -> gaToggleStyleEdit

**Comment by ahocevar** [comment]
I'm still wondering why this test page performs so badly. I modified the official OpenLayers example to use a more resource demanding style than the one in this test page, and still it performs much smoother. See https://69-4974939-gh.circle-artifacts.com/0/examples/mapbox-style.html.

**Comment by davidoesch** [comment]
@ahocevar actually on iPhone latest iOS chrome/safari: 
https://69-4974939-gh.circle-artifacts.com/0/examples/mapbox-style.htm
Crashes/freezes as soon as you zoom in to Europe... Same behavior we have in the viewer

**Comment by ahocevar** [comment]
@davidoesch That's expected because I did not set a smaller tile cache size. See openlayers/openlayers#8956. BTW, that new Safari limitation is a pain. Maybe swisstopo could influence Apple to get rid of it? Reasoning could be that they do not limit the memory used by preloaded images, which is basically the same.

**Comment by davidoesch** [comment]
@boecklic and @pakb : can you take over here and point @ahocevar to the latest code regarding our vec implementation ?
regarding safari limitation: it also is the case for chrom on iOS  --- and we did have some improvements using a small cachesize   but still: performance is far from acceptable -- data issue? ol issue?

**Comment by ahocevar** [comment]
@davidoesch Chrome on iOS is just a wrapper around Safari to add the UI and Google Services. You cannot run any other browser than Safari on iOS.

**Comment by loicgasser** [comment]
Note that the performance might also improve when the tilesets are combined.

**Comment by pakb** [comment]
Here's a codepen comparing mapbox render with OL6 render : https://codepen.io/pakb/full/rPOgYZ
We still have some differences with labels with large zoom levels. Even though we cut Swiss OSM labels from tiles, they're still rendered by OL6 (@gjn suggested it might be through neighbouring tiles that this data could come back). Any hint @ahocevar ?

**Comment by ahocevar** [comment]
@pakb Make sure that all labels are in the same source and are styled/rendererd in the same `ol/layer/VectorTile` instance. Currently it seems that OSM and swisstopo labels are on different sources and rendered on different layers.

**Comment by pakb** [comment]
@ahocevar That's intended, we can't mix our data with OSM because we can't abide to the licence OSM is carrying.
What I could do is set the same zIndexes for our label layers and OSM label layers, but currently this will double the labels as OpenLayers is drawing labels that were specifically cut out of the tiles over Switzerland by our tile server (we are trying to figure out a way to make OL don't draw those labels)

**Comment by ahocevar** [comment]
@pakb With a custom loader, you could combine the features from OSM and swisstopo in a single `ol/source/VectorTile`. Then they would run through the same declutter pass on the same layer.

**Comment by danduk82** [comment]
I'm closing this PR. We have a new branch `mvt_clean` which will be the new *master* branch for vectortiles

**Pull request closed without merging**